File: cpp/expert_trading_features/03_false_sharing.cpp
// ============================================================================
// 03 — False sharing: the silent latency killer (and how alignas kills it)
// Compile:  clang++ -std=c++20 -O2 03_false_sharing.cpp -o bin/03 && bin/03
// ----------------------------------------------------------------------------
// Two threads writing two DIFFERENT variables should be independent. But if both
// variables sit on the SAME 64-byte cache line, every write by one core
// invalidates the line in the other core's cache => the cache-coherence protocol
// (MESI) ping-pongs the line back and forth. Throughput collapses.
//
// Fix: put each hot, independently-written counter on its OWN cache line with
// `alignas(64)`. This is everywhere in trading code: per-thread stats, the
// head/tail of an SPSC queue (see demo 02), per-symbol sequence numbers, etc.
//
// This program measures both layouts so you can see the difference live.
// ============================================================================

#include <atomic>
#include <thread>
#include <chrono>
#include <cstdint>
#include <iostream>
#include <format>

inline constexpr std::size_t kCacheLine = 64;
constexpr std::uint64_t kIters = 100'000'000;

// PACKED: two counters adjacent -> almost certainly same cache line.
struct Packed {
    std::atomic<std::uint64_t> a{0};
    std::atomic<std::uint64_t> b{0};
};

// PADDED: each counter forced onto its own cache line.
struct Padded {
    alignas(kCacheLine) std::atomic<std::uint64_t> a{0};
    alignas(kCacheLine) std::atomic<std::uint64_t> b{0};
};

template <typename Counters>
double bench(const char* label) {
    Counters c;
    auto hammer = [](std::atomic<std::uint64_t>& x) {
        for (std::uint64_t i = 0; i < kIters; ++i)
            x.fetch_add(1, std::memory_order_relaxed);
    };

    auto t0 = std::chrono::steady_clock::now();
    std::thread t1([&] { hammer(c.a); });
    std::thread t2([&] { hammer(c.b); });
    t1.join(); t2.join();
    auto t1d = std::chrono::steady_clock::now();

    double ns = std::chrono::duration<double, std::nano>(t1d - t0).count();
    double per_op = ns / (2.0 * kIters);
    std::cout << std::format("{:<8} {:>6.2f} ns/op   (a={}, b={})\n",
                             label, per_op, c.a.load(), c.b.load());
    return per_op;
}

int main() {
    std::cout << std::format("cache line = {} bytes, {} increments per thread\n\n",
                             kCacheLine, kIters);
    double packed = bench<Packed>("packed");   // false sharing
    double padded = bench<Padded>("padded");   // isolated lines
    std::cout << std::format("\npadding speedup: {:.2f}x  (sizeof Packed={}, Padded={})\n",
                             packed / padded, sizeof(Packed), sizeof(Padded));
    return 0;
}

File: rust/expert_trading_features/examples/03_false_sharing.rs
// ============================================================================
// 03 — False sharing: the silent latency killer (and how alignment kills it)
// Run:  cargo run --release --example 03_false_sharing
// C++ counterpart: cpp/expert_trading_features/03_false_sharing.cpp
// ----------------------------------------------------------------------------
// Two threads writing two DIFFERENT variables should be independent. But if both
// sit on the SAME 64-byte cache line, every write by one core invalidates the
// line in the other's cache => the coherence protocol ping-pongs the line and
// throughput collapses.
//
// Fix: put each hot, independently-written counter on its OWN cache line.
//
// What Rust changes vs C++:
//   * `alignas(64)` becomes `#[repr(align(64))]` (here via the crate's
//     `CachePadded<T>`). Same idea, same effect.
//   * Layout is explicit and the wrapper is a normal type you can reuse — the
//     SPSC ring in demo 02 uses the very same `CachePadded` for its head/tail.
// ============================================================================

use expert_trading_features::CachePadded;
use std::sync::atomic::{AtomicU64, Ordering};
use std::time::Instant;

const ITERS: u64 = 100_000_000;

// PACKED: two counters adjacent -> almost certainly the same cache line.
#[derive(Default)]
struct Packed {
    a: AtomicU64,
    b: AtomicU64,
}

// PADDED: each counter forced onto its own cache line.
#[derive(Default)]
struct PaddedCounters {
    a: CachePadded<AtomicU64>,
    b: CachePadded<AtomicU64>,
}

fn bench(label: &str, a: &AtomicU64, b: &AtomicU64) -> f64 {
    let hammer = |x: &AtomicU64| {
        for _ in 0..ITERS {
            x.fetch_add(1, Ordering::Relaxed);
        }
    };
    let t0 = Instant::now();
    std::thread::scope(|s| {
        s.spawn(|| hammer(a));
        s.spawn(|| hammer(b));
    });
    let ns = t0.elapsed().as_secs_f64() * 1e9;
    let per_op = ns / (2.0 * ITERS as f64);
    println!(
        "{label:<8} {per_op:>6.2} ns/op   (a={}, b={})",
        a.load(Ordering::Relaxed),
        b.load(Ordering::Relaxed)
    );
    per_op
}

fn main() {
    println!("cache line = 64 bytes, {ITERS} increments per thread\n");
    let packed = Packed::default();
    let padded = PaddedCounters::default();

    let p = bench("packed", &packed.a, &packed.b); // false sharing
    let q = bench("padded", &padded.a, &padded.b); // isolated lines

    println!(
        "\npadding speedup: {:.2}x  (sizeof Packed={}, Padded={})",
        p / q,
        std::mem::size_of::<Packed>(),
        std::mem::size_of::<PaddedCounters>()
    );
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn padded_isolates_onto_separate_lines() {
        // Each CachePadded field is 64-byte aligned, so the two counters cannot
        // share a line; the struct is at least two lines wide.
        assert!(std::mem::size_of::<PaddedCounters>() >= 128);
        assert_eq!(std::mem::align_of::<CachePadded<AtomicU64>>(), 64);
    }

    #[test]
    fn packed_is_compact() {
        assert_eq!(std::mem::size_of::<Packed>(), 16); // two u64s, same line
    }
}
